private-identity-audit/wasm/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+        proofs: base64::encode(results.1.encode()),
+    })
+}
+
+/// The second leg of the protocol from PUIS to CDD Provider.
+#[wasm_bindgen]
+pub fn generate_committed_set_and_challenge(
+    private_uuids: Base64,
+    min_set_size: Option<usize>,
+) -> Fallible<CommittedSetOutput> {
+    let uuids: PrivateUids = decode(private_uuids)?;
+
+    let mut rng = OsRng;
+    let results =
+        VerifierSetGenerator::generate_committed_set_and_challenge(uuids, min_set_size, &mut rng)
+            .map_err(|_| WasmError::GenerateCommittedSetError)?;
+
+    Ok(CommittedSetOutput {
+        verifier_secrets: base64::encode(results.0.encode()),
+        committed_uids: base64::encode(results.1.encode()),
+        challenge: base64::encode(results.2.encode()),
+    })
+}
+
+/// The third leg of the protocol from CDD Provider to PUIS.
+#[wasm_bindgen]
+pub fn generate_challenge_response(
+    secrets: Base64,
+    committed_uids: Base64,
+    challenge: Base64,
+) -> Fallible<ChallengeResponseOutput> {
+    let secrets: ProverSecrets = decode(secrets)?;
+    let committed_uids: CommittedUids = decode(committed_uids)?;
+    let challenge: Challenge = decode(challenge)?;
+
+    let mut rng = OsRng;
+    let results =
+        FinalProver::generate_challenge_response(&secrets, &committed_uids, &challenge, &mut rng)
+            .map_err(|_| WasmError::GenerateChallengeResponseError)?;
+
+    Ok(ChallengeResponseOutput {
+        response: base64::encode(results.0.encode()),
+        committed_uids: base64::encode(results.1.encode()),
+    })
+}
+
+/// The last step of the protocol in which PUIS verifies the proofs.
+#[wasm_bindgen]
+pub fn verify_proofs(
+    initial_message: Base64,
+    final_response: Base64,
+    challenge: Base64,
+    cdd_id: Base64,
+    verifier_secrets: Base64,
+    re_committed_uids: Base64,
+) -> Fallible<()> {
+    let initial_message: Proofs = decode(initial_message)?;
+    let final_response: ProverFinalResponse = decode(final_response)?;
+    let challenge: Challenge = decode(challenge)?;
+    let cdd_id: CddId = decode(cdd_id)?;
+    let verifier_secrets: VerifierSecrets = decode(verifier_secrets)?;
+    let re_committed_uids: CommittedUids = decode(re_committed_uids)?;
+
+    Verifier::verify_proofs(
+        &initial_message,
+        &final_response,
+        &challenge,
+        &cdd_id,
+        &verifier_secrets,
+        &re_committed_uids,
+    )
+    .map_err(|_| WasmError::ProofVerificationError)?;
+
+    Ok(())
+}
+
+// ------------------------------------------------------------------------------------
+// -                               Internal Functions                                 -
+// ------------------------------------------------------------------------------------
+
+fn decode<T: Decode>(data: Base64) -> Fallible<T> {
+    let decoded = base64::decode(data).map_err(|_| WasmError::Base64DecodingError)?;
+    T::decode(&mut &decoded[..]).map_err(|_| WasmError::DeserializationError.into())
+}
```

Comment:
This function needs to be at least renamed to disambiguate with `Decode::decode`. I have a vague memory of suggesting `TryFrom` already but probably it was in a different module. Could the following work?
```suggestion
impl TryFrom<Base64> for T where 
    T: Decode
{
    type Error = WasmError;

    fn try_from(data: Base64) -> Result<Self, Self::Error> {
        let decoded = base64::decode(data)
            .map_err(|_| WasmError::Base64DecodingError)?;
        T::decode(&mut &decoded[..]).map_err(|_| WasmError::DeserializationError)
    }
}
```

Reply:
I tried this and it indeed failed because `Base64: Decode`.

So perhaps simply renaming this `decode` function as something like `decode_base64` would help differentiating between various steps of the decoding process.